File: lib/domain/entities/bike.dart
import 'package:equatable/equatable.dart';

class BikeSpecifications extends Equatable {
  final String engine;
  final String power;
  final String torque;
  final String transmission;
  final String weight;
  final String fuelCapacity;

  const BikeSpecifications({
    required this.engine,
    required this.power,
    required this.torque,
    required this.transmission,
    required this.weight,
    required this.fuelCapacity,
  });

  @override
  List<Object?> get props => [
        engine,
        power,
        torque,
        transmission,
        weight,
        fuelCapacity,
      ];
}

class Review extends Equatable {
  final String userId;
  final String userName;
  final int rating;
  final String comment;
  final DateTime date;

  const Review({
    required this.userId,
    required this.userName,
    required this.rating,
    required this.comment,
    required this.date,
  });

  @override
  List<Object?> get props => [
        userId,
        userName,
        rating,
        comment,
        date,
      ];
}

class Bike extends Equatable {
  final String id;
  final String name;
  final String brand;
  final double price;
  final String category;
  final String imageUrl;
  final String description;
  final BikeSpecifications specifications;
  final int stock;
  final bool isFeatured;
  final double rating;
  final List<Review> reviews;

  const Bike({
    required this.id,
    required this.name,
    required this.brand,
    required this.price,
    required this.category,
    required this.imageUrl,
    required this.description,
    required this.specifications,
    required this.stock,
    required this.isFeatured,
    required this.rating,
    required this.reviews,
  });

  @override
  List<Object?> get props => [
        id,
        name,
        brand,
        price,
        category,
        imageUrl,
        description,
        specifications,
        stock,
        isFeatured,
        rating,
        reviews,
      ];
} 

File: backend/src/models/Bike.js
const mongoose = require('mongoose');

const reviewSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  userName: {
    type: String,
    required: true
  },
  rating: {
    type: Number,
    required: true,
    min: 1,
    max: 5
  },
  comment: {
    type: String,
    required: true
  },
  date: {
    type: Date,
    default: Date.now
  }
});

const specificationsSchema = new mongoose.Schema({
  engine: {
    type: String,
    required: true
  },
  power: {
    type: String,
    required: true
  },
  torque: {
    type: String,
    required: true
  },
  transmission: {
    type: String,
    required: true
  },
  weight: {
    type: String,
    required: true
  },
  fuelCapacity: {
    type: String,
    required: true
  }
});

const bikeSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true
  },
  brand: {
    type: String,
    required: true,
    trim: true
  },
  price: {
    type: Number,
    required: true,
    min: 0
  },
  category: {
    type: String,
    required: true,
    enum: ['Sport', 'Cruiser', 'Adventure', 'Naked', 'Touring']
  },
  imageUrl: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  specifications: {
    type: specificationsSchema,
    required: true
  },
  stock: {
    type: Number,
    required: true,
    min: 0
  },
  isFeatured: {
    type: Boolean,
    default: false
  },
  rating: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },
  reviews: [reviewSchema]
}, {
  timestamps: true
});

module.exports = mongoose.model('Bike', bikeSchema); 